File: app/src/main/java/com/example/exoesqueletov1/fragments/PairedDevisesFragment.java
package com.example.exoesqueletov1.fragments;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;

import com.example.exoesqueletov1.ControlActivity;
import com.example.exoesqueletov1.R;

import java.util.ArrayList;
import java.util.Set;

public class PairedDevisesFragment extends Fragment {

    private String typeUser;
    private BluetoothAdapter bluetoothAdapter = null;
    public static final String DEVICE_ADDRESS = "deviceAddress";
    public static final String TYPE_USER = "typeUser";
    private static final int REQUEST_CODE = 1;

    private ListView listView;


    public PairedDevisesFragment(String typeUser) {
        this.typeUser = typeUser;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_paired_devises, container, false);

        Button buttonPaired = view.findViewById(R.id.button_paired_devises);
        listView = view.findViewById(R.id.listView);

        bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();

        if (bluetoothAdapter == null) {
            Toast.makeText(getContext(), R.string.no_bluetooth, Toast.LENGTH_LONG).show();
            getActivity().finish();
        }
        if (!bluetoothAdapter.isEnabled()) {
            Intent turnBTon = new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE);
            startActivityForResult(turnBTon, REQUEST_CODE);
        }

        pairedDevicesList();

        buttonPaired.setOnClickListener(v -> pairedDevicesList());

        return view;
    }

    private void pairedDevicesList() {
        Set<BluetoothDevice> bluetoothDevices = bluetoothAdapter.getBondedDevices();
        ArrayList list = new ArrayList();

        if (bluetoothDevices.size() > 0) {
            for (BluetoothDevice bt : bluetoothDevices) {
                list.add(bt.getName() + "\n" + bt.getAddress());
                //Get the device's name and the address
            }
        } else {
            Toast.makeText(getContext(), "No hay dispositivos",
                    Toast.LENGTH_LONG).show();
        }

        final ArrayAdapter adapter;
        adapter = new ArrayAdapter(getContext(), R.layout.list_item_withe, R.id.list_content, list);
        listView.setAdapter(adapter);
        listView.setOnItemClickListener(myListClickListener);
    }

    private AdapterView.OnItemClickListener myListClickListener = (av, view, arg2, arg3) -> {
        TextView textView = view.findViewById(R.id.list_content);
        String info = textView.getText().toString();
        String address = info.substring(info.length() - 17);
        Intent intent = new Intent(getContext(), ControlActivity.class);
        intent.putExtra(DEVICE_ADDRESS, address);
        intent.putExtra(TYPE_USER, typeUser);
        startActivity(intent);
    };
}

File: app/src/main/java/com/example/exoesqueletov1/ui/fragments/pairedDevises/PairedDevisesFragment.kt
package com.example.exoesqueletov1.ui.fragments.pairedDevises

import android.Manifest
import android.annotation.SuppressLint
import android.app.Activity
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothManager
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.activity.result.contract.ActivityResultContracts
import androidx.core.app.ActivityCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.exoesqueletov1.R
import com.example.exoesqueletov1.data.local.query.ExoskeletonQuery
import com.example.exoesqueletov1.databinding.FragmentPairedDevisesBinding
import com.example.exoesqueletov1.interfaces.BluetoothResource
import com.example.exoesqueletov1.service.model.BluetoothResponse
import com.example.exoesqueletov1.ui.activity.main.MainActivity
import com.example.exoesqueletov1.ui.fragments.pairedDevises.adapter.PairedDevicesAdapter
import com.example.exoesqueletov1.utils.Constants
import dagger.hilt.android.AndroidEntryPoint
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

@AndroidEntryPoint
class PairedDevisesFragment : Fragment() {

    private lateinit var binding: FragmentPairedDevisesBinding
    private var bluetoothAdapter: BluetoothAdapter? = null
    private var bondedDevices: MutableSet<BluetoothDevice>? = null
    private lateinit var viewModel: PairedDeviceViewModel
    private val listExoskeletonQuery = mutableListOf<ExoskeletonQuery>()
    private val listDevices = mutableListOf<ExoskeletonQuery>()
    private lateinit var adapter: PairedDevicesAdapter

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        viewModel = ViewModelProvider(this)[PairedDeviceViewModel::class.java]
        binding = FragmentPairedDevisesBinding.inflate(layoutInflater, container, false)
        return binding.root
    }

    @SuppressLint("ServiceCast")
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        adapter = PairedDevicesAdapter(listDevices) {
            binding.textStatus.text = "Conectando..."
        }
        bluetoothAdapter =
            (requireContext().getSystemService(Context.BLUETOOTH_SERVICE) as BluetoothManager).adapter

        viewModel.bluetoothDevice.observe(viewLifecycleOwner) {
            listExoskeletonQuery.clear()
            listExoskeletonQuery.addAll(it)
        }

        if (ActivityCompat.checkSelfPermission(
                requireContext(),
                Manifest.permission.BLUETOOTH_CONNECT
            ) == PackageManager.PERMISSION_GRANTED
        ) {
            if (bluetoothAdapter != null) {
                if (!bluetoothAdapter!!.isEnabled) {
                    binding.textStatus.text = getString(R.string.bluetooth_inavil)
                    val enableIntent = Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)
                    isEnableBluetooth.launch(enableIntent)
                } else setListeners()
            } else binding.textStatus.text = getString(R.string.bluetooth_not_found)
        } else binding.textStatus.text = getString(R.string.bluetooth_permission)

        binding.apply {
            recyclerBluetoothDevices.setHasFixedSize(true)
            recyclerBluetoothDevices.layoutManager =
                LinearLayoutManager(requireContext(), LinearLayoutManager.VERTICAL, false)
            recyclerBluetoothDevices.adapter = adapter
        }
    }

    @SuppressLint("MissingPermission", "NotifyDataSetChanged")
    private fun startProcess() {
        if (listDevices.isNotEmpty() && (listDevices.size == 1) && bluetoothAdapter!!.isDiscovering) {
            EventBus.getDefault().post(listDevices[0])
            binding.textStatus.text = "Conectando..."
            binding.materialCardView3.visibility = View.GONE
            bluetoothAdapter!!.cancelDiscovery()
            requireActivity().unregisterReceiver(broadcastReceiver)
            return
        }
        listDevices.clear()
        bondedDevices = bluetoothAdapter!!.bondedDevices
        bondedDevices!!.forEach { device ->
            listExoskeletonQuery.forEach { exoskeletonModel ->
                if (device.address == exoskeletonModel.mac) {
                    exoskeletonModel.status = Constants.StatusDevice.Emparejado.name
                    exoskeletonModel.name = device.name
                    listDevices.add(exoskeletonModel)
                }
            }
        }
        if (!bluetoothAdapter!!.isDiscovering) {
            val filter = IntentFilter()
            filter.addAction(BluetoothDevice.ACTION_FOUND)
            filter.addAction(BluetoothAdapter.ACTION_DISCOVERY_FINISHED)
            filter.addAction(BluetoothAdapter.ACTION_DISCOVERY_STARTED)
            requireActivity().registerReceiver(broadcastReceiver, filter)
            bluetoothAdapter!!.startDiscovery()
        } else {
            binding.textStatus.text = getString(R.string.listo)
            bluetoothAdapter!!.cancelDiscovery()
            requireActivity().unregisterReceiver(broadcastReceiver)
        }
        if (listDevices.isNotEmpty()) adapter.notifyDataSetChanged()

    }

    private fun setListeners() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            binding.textStatus.text = getString(R.string.listo)
            binding.imageView9.setOnClickListener {
                startProcess()
            }
        }
    }

    private val broadcastReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        @SuppressLint("MissingPermission", "NotifyDataSetChanged")
        override fun onReceive(context: Context, intent: Intent) {
            when (intent.action) {
                BluetoothDevice.ACTION_FOUND -> {
                    if (listDevices.size == 1) binding.materialCardView3.visibility = View.GONE
                    if (listDevices.size > 1) binding.materialCardView3.visibility = View.VISIBLE
                    val device =
                        intent.getParcelableExtra<BluetoothDevice>(BluetoothDevice.EXTRA_DEVICE)
                    listExoskeletonQuery.forEach { exoskeletonModel ->
                        if (device!!.address == exoskeletonModel.mac) {
                            exoskeletonModel.status = Constants.StatusDevice.Cercano.name
                            exoskeletonModel.name = device.name
                            listDevices.add(exoskeletonModel)
                        }
                    }
                    if (listDevices.isNotEmpty()) adapter.notifyDataSetChanged()
                }
                BluetoothAdapter.ACTION_DISCOVERY_STARTED -> {
                    binding.textStatus.text = "Da click para cancelar"
                }
                BluetoothAdapter.ACTION_DISCOVERY_FINISHED -> {
                    binding.textStatus.text = getString(R.string.reintentar)
                    if (listDevices.isNotEmpty() && listDevices.size == 1) {
                        EventBus.getDefault().post(listDevices[0])
                        binding.textStatus.text = "Conectando..."
                        binding.recyclerBluetoothDevices.visibility = View.GONE
                    }
                }
            }
        }
    }

    @SuppressLint("MissingPermission")
    private var isEnableBluetooth =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
            if (it.resultCode == Activity.RESULT_OK) {
                setListeners()
            }
        }

    /**
     * Method used by **get Status** and data of bluetooth connection in
     * [MainActivity.onSerialConnect], [MainActivity.onSerialConnectError] ...
     * @param resource Is a object type [BluetoothResource] and contains the status
     * in a object type [Constants.StatusBluetoothDevice] and Data in object type [BluetoothResponse].
     *
     * **[BluetoothResource.status] just receive:**
     * - [Constants.StatusBluetoothDevice.Connected]
     * - [Constants.StatusBluetoothDevice.ConnectionFailed]
     *
     */
    @Subscribe(threadMode = ThreadMode.MAIN)
    fun getStatus(resource: BluetoothResource) {
        when (resource.status) {
            Constants.StatusBluetoothDevice.Connected -> {
                //findNavController().navigate(R.id.action_navigation_paired_device_to_connectionFragment)
            }
            Constants.StatusBluetoothDevice.ConnectionFailed -> {
                binding.textStatus.text =
                    "No se pudo conectar con el exoeskeleto.\nIntente otra vez."
                if (listDevices.size > 1) binding.recyclerBluetoothDevices.visibility = View.VISIBLE
            }
            else -> {}
        }
    }

    override fun onStart() {
        super.onStart()
        EventBus.getDefault().register(this)
    }

    override fun onPause() {
        super.onPause()
        EventBus.getDefault().unregister(this)
    }

}
